cognite/client/data_classes/data_modeling/typed_instances.py

```diff
@@ -0,0 +1,172 @@
+        raise NotImplementedError()
+
+    def dump(self, camel_case: bool = True) -> dict[str, str | dict]:
+        output = self._dump_instance(camel_case)
+        properties = self._dump_properties(camel_case)
+        if properties:
+            output["sources"] = [
+                {
+                    "source": self.get_source().dump(camel_case),
+                    "properties": properties,
+                }
+            ]
+        return output
+
+    def _dump_instance(self, camel_case: bool) -> dict[str, Any]:
+        output: dict[str, Any] = {
+            "space": self.space,
+            "externalId" if camel_case else "external_id": self.external_id,
+            "instanceType" if camel_case else "instance_type": self._instance_type,
+        }
+        if self.existing_version:
+            output["existingVersion" if camel_case else "existing_version"] = self.existing_version
+        return output
+
+    def _dump_properties(self, camel_case: bool) -> dict[str, Any]:
+        properties: dict[str, str | int | float | bool | dict | list] = {}
+        for key, value in vars(self).items():
+            if key in self._instance_properties:
+                continue
+            if isinstance(value, Iterable) and not isinstance(value, (str, dict)):
+                properties[key] = [self._serialize_value(v, camel_case) for v in value]
+            else:
+                properties[key] = self._serialize_value(value, camel_case)
+        return properties
+
+    @staticmethod
+    def _serialize_value(value: PropertyValueWrite, camel_case: bool) -> str | int | float | bool | dict | list:
+        if isinstance(value, (NodeId, EdgeId)):
+            # We don't want to dump the instance_type field when serializing NodeId in this context
+            return value.dump(camel_case, include_instance_type=False)
+        elif isinstance(value, DirectRelationReference):
+            return value.dump(camel_case)
+        elif isinstance(value, datetime):
+            return value.isoformat(timespec="milliseconds")
+        elif isinstance(value, date):
+            return value.isoformat()
+        else:
+            return value
+
+    @classmethod
+    def _load(cls, resource: dict[str, Any], cognite_client: CogniteClient | None = None) -> Self:
+        args: dict[str, Any] = {}
+        args.update(cls._load_properties(resource))
+        args.update(cls._load_instance(resource))
+        return cls(**args)
+
+    @classmethod
+    def _load_instance(cls, resource: dict[str, Any]) -> dict[str, Any]:
+        args: dict[str, Any] = {}
+        for key in cls._instance_properties:
+            camel_key = to_camel_case(key)
+            if camel_key in resource:
+                args[key] = resource[camel_key]
+        return args
+
+    @classmethod
+    def _load_properties(cls, resource: dict[str, Any]) -> dict[str, Any]:
+        raise NotImplementedError()
+
+
+class TypedNodeWrite(TypedInstanceWrite, ABC):
```

Comment:
Would be good with some unit tests for a subclass of this, since we won't get any generated when these are abstract